echopype/convert/set_groups_base.py

```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+        to copy the array (hence the .copy()). This allows the array
+        to be writable downstream (i.e. we can assign values to
+        certain indices).
+
+        To retain the attributes and encoding of ``beam``
+        it is necessary to use .assign_coords() with ``beam``
+        from ds.
+        """
+
+        # variables to add beam to
+        add_beam_names = set(ds.variables).intersection(
+            beam_only_names[sonar_model].union(beam_ping_time_names[sonar_model])
+        )
+
+        for var_name in add_beam_names:
+            if "beam" in ds.dims:
+                if "beam" not in ds[var_name].dims:
+                    ds[var_name] = (
+                        ds[var_name]
+                        .expand_dims(dim={"beam": ds.beam}, axis=ds[var_name].ndim)
+                        .assign_coords(beam=ds.beam)
+                        .copy()
+                    )
+            else:
+                # TODO: right now there is no attr or encoding for the beam dimension
+                #  if this changes in the future, we need to add them here.
+                ds[var_name] = (
+                    ds[var_name]
+                    .expand_dims(dim={"beam": np.array(["1"], dtype=str)}, axis=ds[var_name].ndim)
+                    .copy()
+                )
+
+    @staticmethod
+    def _add_ping_time_dim(ds: xr.Dataset, sonar_model: str):
+        """
+        Adds ``ping_time`` as the last dimension to the appropriate
+        variables in ``Sonar/Beam_group1`` and ``Sonar/Beam_group2``
+        (when necessary).
+
+        Notes
+        -----
+        When expanding the dimension of a Dataarray, it is necessary
+        to copy the array (hence the .copy()). This allows the array
+        to be writable downstream (i.e. we can assign values to
+        certain indices).
+
+        To retain the attributes and encoding of ``ping_time``
+        it is necessary to use .assign_coords() with ``ping_time``
+        from ds.
+        """
+
+        # variables to add ping_time to
+        add_ping_time_names = set(ds.variables).intersection(beam_ping_time_names[sonar_model])
+        add_ping_time_names = add_ping_time_names.union(ping_time_only_names[sonar_model])
+
+        ping_time = ds.ping_time.values
+
+        for var_name in add_ping_time_names:
+
+            ds[var_name] = (
+                ds[var_name]
+                .expand_dims(dim={"ping_time": ping_time}, axis=ds[var_name].ndim)
+                .assign_coords(ping_time=ds.ping_time)
+                .copy()
+            )
+
+    def beamgroups_to_convention(self, ds: xr.Dataset, sonar_model: str):
+        """
+        Manipulates the variables in ``Sonar/Beam_group1``
+        and ``Sonar/Beam_group2`` (when necessary) so that
+        they adhere to convention.
+
+        This does several things:
+        1. Adds ``beam`` dimension to several variables
+        2. Adds ``ping_time`` dimension to several variables
+        """
```

Comment:
Some suggested edits to clarify what's done and the context (note: I don't know if some of the new line lengths exceed a docstring requirement):

```
Manipulates variables in ``Sonar/Beam_group1``
and ``Sonar/Beam_group2`` to adhere to 
SONAR-netCDF4 vers. 1 with respect to the use of
``ping_time`` and ``beam`` dimensions.

This does several things:
1. Creates ``beam`` dimension and coordinate variable when not present
2. Adds ``beam`` dimension to several variables when missing
3. Adds ``ping_time`` dimension to several variables when missing
```